src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs

```diff
@@ -9,6 +10,49 @@
 
 namespace Microsoft.CodeAnalysis.Shared.Extensions
 {
+    internal struct TokenSemanticInfo
+    {
+        public static readonly TokenSemanticInfo Empty = new TokenSemanticInfo(
+            null, null, ImmutableArray<ISymbol>.Empty, null);
+
+        public readonly ISymbol DeclaredSymbol;
+        public readonly IAliasSymbol AliasSymbol;
+        public readonly ImmutableArray<ISymbol> ReferencedSymbols;
+        public readonly ITypeSymbol Type;
+
+        public TokenSemanticInfo(
+            ISymbol declaredSymbol, 
+            IAliasSymbol aliasSymbol,
+            ImmutableArray<ISymbol> referencedSymbols,
+            ITypeSymbol type)
+        {
+            DeclaredSymbol = declaredSymbol;
+            AliasSymbol = aliasSymbol;
+            ReferencedSymbols = referencedSymbols;
+            Type = type;
+        }
+
+        public ImmutableArray<ISymbol> GetSymbols(bool includeType)
+        {
+            var result = ArrayBuilder<ISymbol>.GetInstance();
+            result.AddIfNotNull(DeclaredSymbol);
+            result.AddIfNotNull(AliasSymbol);
+            result.AddRange(ReferencedSymbols);
+
+            if (includeType)
+            {
+                result.AddIfNotNull(Type);
+            }
+
+            return result.ToImmutableAndFree();
+        }
+
+        public ISymbol GetAnySymbol(bool includeType)
+        {
+            return GetSymbols(includeType).FirstOrDefault();
```

Comment:
Seems like you could inline the order here and avoid ever copying into the immutable array.